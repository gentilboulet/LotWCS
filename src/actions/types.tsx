import { IInitialStateAction } from '../actions/initial';
import { IHeaderAction } from '../actions/header';
import { IChiAction } from '../actions/chi';
import { ISkillAction } from '../actions/skills';
import { ILoresheetAction } from '../actions/loresheets';
import { IHistoryAction } from '../actions/history';
import { IKungFuAction } from '../actions/kungfus';
import { IVirtueAction } from '../actions/virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
