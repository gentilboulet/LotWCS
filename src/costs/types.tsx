export interface ICost {
  destiny: number;
  discountIdx: number;
  discountNewValue: number;
  entanglement: number;
}

export const defaultCost: ICost = {
  destiny: 0,
  discountIdx: -1,
  discountNewValue: 0,
  entanglement: 0,
};
