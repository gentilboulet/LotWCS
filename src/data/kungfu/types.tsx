import { TChiName } from 'data/chi';
import { WEAPON_TYPE } from 'data/weapons';
import { IEffect } from 'state/effects';

// Constants used to differentiate between kungfu
export const KUNGFU_EXTERNAL = 'KUNGFU_EXTERNAL';
export type KUNGFU_EXTERNAL = typeof KUNGFU_EXTERNAL;

export const KUNGFU_INTERNAL = 'KUNGFU_INTERNAL';
export type KUNGFU_INTERNAL = typeof KUNGFU_INTERNAL;

export type KUNGFU_TYPE = KUNGFU_INTERNAL | KUNGFU_EXTERNAL;

export interface IDataExternalKungfuStatistics {
    speed: 0|5|10|15|20;
    footwork: 0|5|10|15|20;
    strike: 0|5|10|15|20;
    damage: 0|5|10|15|20;
    block: 0|5|10|15|20;
    toughness: 0|5|10|15|20;
}

export interface IDataExternalKungfuPrerequisite { uid: string };

export interface IDataExternalKungfuTechnique {
    name: string;
    cost: number;
    description: string;
    effect: IEffect;
    uid: string;
    prerequisites?: IDataExternalKungfuPrerequisite[];
}

export interface IDataInternalKungfuTechnique {
    name: string;
    level: number; // 1 | 2 | 3 | 4 | 5;
    description: string;
    effect: IEffect;
    uid: string;
}

export interface IDataExternalKungfu {
  name: string;
  laugths: string;
  fears: string;
  weapons: WEAPON_TYPE[];
  statistics: IDataExternalKungfuStatistics;
  techniques: IDataExternalKungfuTechnique[];
  uid: string;
}

export interface IDataInternalKungfu {
  name: string;
  element: TChiName;
  techniques: IDataInternalKungfuTechnique[];
  uid: string;
}
