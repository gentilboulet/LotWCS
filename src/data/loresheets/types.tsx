export type SECRET_ARTS = typeof SECRET_ARTS;
export const SECRET_ARTS = "SECRET_ARTS";

export type SAGE_LORESHEET = typeof SAGE_LORESHEET;
export const SAGE_LORESHEET = "SAGE_LORESHEET";
