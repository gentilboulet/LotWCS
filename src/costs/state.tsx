import { ICost } from 'costs/types';
import { IDiscount } from 'perks/types/discounts';
import { IStoreState } from 'state/types';

import * as constants from 'perks/constants/discounts';
import * as dataLoresheets from 'data/loresheets';
import * as loresheets from 'state/loresheets';
import * as skills from 'state/skills';

import * as derived from 'state/derived';

// helper functions
function _canHandleCost(state: IStoreState, cost: ICost): boolean {
  if ( (state.get('destiny') - cost.destiny) < 0 ) { return false; }
  if ( (state.get('entanglement') - cost.entanglement) < 0 ) { return false; }
  return true;
}

function _handleDiscount(state: IStoreState, idx: number, cost: number): ICost {
  if (idx === -1) {
    return { destiny: cost, discountIdx: idx, discountNewValue: 0, entanglement: 0, };
  } else {
    const discount = state.getIn(['discounts', idx]).value;
    const remainingCost = Math.max(0, cost - discount);
    const usedDiscountValue = cost - remainingCost;
    return {
      destiny: remainingCost,
      discountIdx: idx,
      discountNewValue: discount - usedDiscountValue,
      entanglement: 0,
    };
  }
}

/*
  Export canBuy and getCost for some actions. Available :
    canBuySkill
    getCostSkill
    canBuySpeciality
    getCostSpeciality
    canOpenLoresheet
    getCostOpenLoresheet
    canBuyOptionLoresheet
    getCostsArrayBuyOptionLoresheet
 */

export function canBuySkill(state: IStoreState, skill: string): boolean {
  const value = skills.getSkillValue(state, skill);
  if ( (value + 5) > derived.maxSkillBonus(state) ) { return false; }
  const cost = getCostSkill(state, skill);
  return _canHandleCost(state, cost);
}

export function getCostSkill(state: IStoreState, skill: string): ICost {
  const defCost = 2;

  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_SKILL )
      && (r.skills.findIndex((s: string) => (s === skill) ) >= 0);
    });

  return _handleDiscount(state, idx, defCost);
}

export function canBuySpeciality(state: IStoreState, skill: string, speciality: string): boolean {
  const specialityIdx = skills.getSpecialityIndex(state, skill, speciality);

  if ( specialityIdx < 0 ) { return false; }

  const cost = getCostSpeciality(state, skill, speciality);
  return _canHandleCost(state, cost);
}

export function getCostSpeciality(state: IStoreState, skill: string, speciality: string): ICost {
  const defCost = 1;

  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_SKILL )
        && (r.skills.findIndex((s: string) => (s === skill)) >= 0);
    });

  return _handleDiscount(state, idx, defCost);
}

export function canOpenLoresheet(state: IStoreState, uid: string, openCost: number): boolean {
  const idx = loresheets.getLoresheetIndex(state, uid);
  if (idx !== -1) { return false; } // Already opened
  const cost: ICost = getCostOpenLoresheet(state, uid, openCost);
  return _canHandleCost(state, cost);
}

export function getCostOpenLoresheet(state: IStoreState, uid: string, cost: number): ICost {
  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET )
      && (r.uids.findIndex((id: string) => (id === uid)) >= 0);
    });
  return _handleDiscount(state, idx, cost);
}

export function canBuyOptionLoresheet(state: IStoreState, lsUid: string, uid: string, buyCost: number): boolean {
  if ( loresheets.getLoresheetIndex(state, lsUid) === -1) { return false; } // LS not open

  const dataOpt = dataLoresheets.optionLS(lsUid, uid);
  const idxOpt = loresheets.getLoresheetOptionIndex(state, lsUid, uid);

  if (idxOpt === -1 && !dataOpt.repeatable) { return false; } // Already bought & not repeatable

  if (dataOpt.prerequisites.filter( // filter checked prereqs
      (p: dataLoresheets.IDataLoresheetOptionPrerequisite) => {
        if (typeof p === 'string') {
          return loresheets.getLoresheetOptionIndex(state, lsUid, p) !== -1;
        } else if ( p.type === 'OR' ) {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length > 0;
        } else if ( p.type === 'AND') {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length === p.prerequisites.length;
        }
        return false;
      }
    ).length !== dataOpt.prerequisites.length ) { return false; } // prereqs not cleared

  const cost: ICost = getCostBuyOptionLoresheet(state, lsUid, uid, buyCost);
  return _canHandleCost(state, cost);
}

export function getCostBuyOptionLoresheet(state: IStoreState, lsUid: string, uid: string, cost: number): ICost {
  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET_OPTION )
      && (r.uids.findIndex(id => { return id.lsUid === lsUid &&
        id.optUid.findIndex(optUid => (optUid === uid)) >= 0; })
      >= 0);
    });
  return _handleDiscount(state, idx, cost);
}
