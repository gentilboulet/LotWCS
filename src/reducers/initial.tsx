import * as Immutable from 'immutable';
import { makeTypedFactory } from 'typed-immutable-record';

import {
  defaultChiFactory,
  IStoreKungFu, 
  IStoreLoresheet, IStoreLoresheetOption,
  IStoreSkill, IStoreSkillSpeciality,
  IStoreState, IStoreStateJS,
  skillFactory,
} from '../types/state';

import { resetToInitialState } from '../actions/initial';

import { IAction } from '../types/actions';

import { IDataSkill, skills } from '../data/skills';

export const defaultStateJS: IStoreStateJS = {
  name: 'No Name',
  concept: 'No Concept',
  archetype: '',
  archetypeModified: false,
  rank: '',
  rankValue: -1,
  rankModified: false,
  entanglement: 0,
  destiny: 0,
  skills: Immutable.List<IStoreSkill>(skills.map((s: IDataSkill) => {
    return skillFactory({
      name: s.name,
      value: 0,
    });
  })),
  skillSpecialities: Immutable.List<IStoreSkillSpeciality>([]),
  loresheets: Immutable.List<IStoreLoresheet>([]),
  loresheetOptions: Immutable.List<IStoreLoresheetOption>([]),
  discounts: Immutable.List([]),
  bonuses: Immutable.List([]),
  history: Immutable.List<IAction>([resetToInitialState()]),
  chi: defaultChiFactory(),
  externalKungFus: Immutable.List<IStoreKungFu>([]),
  internalKungFus: Immutable.List<IStoreKungFu>([]),
};

export const initialStateFactory = makeTypedFactory<IStoreStateJS, IStoreState>(defaultStateJS);
