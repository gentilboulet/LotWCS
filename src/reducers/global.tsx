import { IAction } from '../types/actions';
import { IStoreState } from '../types/state';

// Sub Reducers
import * as header from '../constants/header';
import { headerReducer } from '../reducers/header';

import * as initial from '../constants/initial';
import { initialStateFactory } from '../reducers/initial';

import * as history from '../constants/history';
import { historyReducer } from '../reducers/history';

import * as skills from '../constants/skills';
import { skillsReducer } from '../reducers/skills';

import * as loresheets from '../constants/loresheets';
import { loresheetsReducer } from '../reducers/loresheets';

export function globalReducer(state: IStoreState, action: IAction): IStoreState {
  switch (action.type) {
    case header.HEADER_SET_ARCHETYPE:
    case header.HEADER_SET_CONCEPT:
    case header.HEADER_SET_NAME:
    case header.HEADER_SET_RANK:
      return headerReducer(state, action);
    case initial.INITIAL_STATE:
      return initialStateFactory();
    case skills.SKILLS_BUY:
      return skillsReducer(state, action);
    case loresheets.LORESHEET_OPEN:
    case loresheets.LORESHEET_BUY_OPTION:
      return loresheetsReducer(state, action);
    case history.HISTORY_DELETE:
      return historyReducer(state, action);
    default:
      return state;
  }
}
