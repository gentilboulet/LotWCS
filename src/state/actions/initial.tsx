import * as constants from 'constants/initial';

export interface IInitialState {
  type: constants.INITIAL_STATE;
}

export type IInitialStateAction = IInitialState;

export function resetToInitialState(): IInitialState {
  return {type: constants.INITIAL_STATE};
}
