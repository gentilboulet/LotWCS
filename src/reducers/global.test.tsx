import { IStoreState } from '../types/state';
import { IAction } from '../types/actions';

import { initialStateFactory } from './initial';
import { globalReducer } from './global';

const initialState: IStoreState  = initialStateFactory();

it('JUNK', () => {
  expect( initialState ).toMatchSnapshot();
  const junk = { type: 'JUNK_ACTION' };
  expect( globalReducer(initialState, junk as IAction )).toMatchObject( initialState );
});
