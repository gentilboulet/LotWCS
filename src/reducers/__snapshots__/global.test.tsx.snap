// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing globalReducer should do nothing with a junk action 1`] = `
Object {
  "archetype": "",
  "archetypeModified": false,
  "bonuses": Array [],
  "chi": Object {
    "demon": 0,
    "demonCultivation": 0,
    "earth": 0,
    "earthCultivation": 0,
    "enlightened": 0,
    "enlightenedCultivation": 0,
    "fire": 0,
    "fireCultivation": 0,
    "general": 0,
    "generalCultivation": 0,
    "metal": 0,
    "metalCultivation": 0,
    "water": 0,
    "waterCultivation": 0,
    "wood": 0,
    "woodCultivation": 0,
  },
  "concept": "No Concept",
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "externalKungFus": Array [],
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
  ],
  "internalKungFus": Array [],
  "loresheetOptions": Array [],
  "loresheets": Array [],
  "name": "No Name",
  "rank": "",
  "rankModified": false,
  "rankValue": -1,
  "skillSpecialities": Array [],
  "skills": Array [
    Object {
      "name": "Awareness",
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "value": 0,
    },
    Object {
      "name": "Learning",
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "value": 0,
    },
    Object {
      "name": "Might",
      "value": 0,
    },
    Object {
      "name": "Perform",
      "value": 0,
    },
    Object {
      "name": "Politics",
      "value": 0,
    },
    Object {
      "name": "Ride",
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "value": 0,
    },
    Object {
      "name": "Survival",
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "value": 0,
    },
  ],
}
`;
